adapters/json/json.go

```diff
@@ -0,0 +1,139 @@
+package json
+
+import (
+	"encoding/json"
+	"errors"
+	"log"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+	"text/template"
+
+	"github.com/gliderlabs/logspout/adapters/syslog"
+	"github.com/gliderlabs/logspout/router"
+)
+
+var configDefaults = map[string]string{
+	"JSON_FIELDS":          "time:uint,message,docker.hostname,docker.image",
+	"JSON_TIME":            "{{.Time.Unix}}",
+	"JSON_MESSAGE":         "{{.Data}}",
+	"JSON_SOURCE":          "{{.Source}}",
+	"JSON_DOCKER_HOSTNAME": "{{.Container.Config.Hostname}}",
+	"JSON_DOCKER_IMAGE":    "{{.Container.Config.Image}}",
+	"JSON_DOCKER_ID":       "{{.Container.ID}}",
+	"JSON_DOCKER_NAME":     "{{.ContainerName}}",
+}
+
+func init() {
+	router.AdapterFactories.Register(NewJSONAdapter, "json")
+}
+
+func getopt(name string) string {
+	value := os.Getenv(name)
+	if value == "" {
+		value = configDefaults[name]
+	}
+	return value
+}
+
+type JSONAdapter struct {
+	conn  net.Conn
+	route *router.Route
+	tmpl  *template.Template
+	types map[string]string
+}
+
+func NewJSONAdapter(route *router.Route) (router.LogAdapter, error) {
+	transport, found := router.AdapterTransports.Lookup(route.AdapterTransport("udp"))
+	if !found {
+		return nil, errors.New("unable to find adapter: " + route.Adapter)
+	}
+	conn, err := transport.Dial(route.Address, route.Options)
+	if err != nil {
+		return nil, err
+	}
+
+	fields := strings.Split(getopt("JSON_FIELDS"), ",")
+	types := make(map[string]string)
+	var values []string
+	for _, field := range fields {
+		parts := strings.Split(field, ":")
+		if len(parts) > 1 {
+			types[parts[0]] = parts[1]
+		}
+		config := "JSON_" + strings.ToUpper(strings.Replace(parts[0], ".", "_", -1))
+		values = append(values, parts[0]+":"+getopt(config))
+	}
+	tmplStr := strings.Join(values, "\x00")
+	tmpl, err := template.New("prejson").Parse(tmplStr)
+	if err != nil {
+		return nil, err
+	}
+
+	return &JSONAdapter{
+		route: route,
+		conn:  conn,
+		tmpl:  tmpl,
+		types: types,
+	}, nil
+}
+
+func (a *JSONAdapter) Stream(logstream chan *router.Message) {
+	defer a.route.Close()
+	for message := range logstream {
+		m := syslog.NewSyslogMessage(message, a.conn)
+		buf, err := m.Render(a.tmpl)
+		if err != nil {
+			log.Println("json:", err)
+			return
+		}
+		data, err := json.Marshal(buildMap(buf.String(), a.types))
+		if err != nil {
+			log.Println("json:", err)
+			return
+		}
+		_, err = a.conn.Write(data)
+		if err != nil {
+			log.Println("json:", err)
+			return
+		}
+	}
+}
+
+func buildMap(input string, types map[string]string) map[string]interface{} {
```

Comment:
No I understand. I was getting caught up in style of communication. I appreciate your feedback and contributions, and any benchmarking around this will be very helpful to document. Not necessarily to prove a difference, but to show what scenario this adapter should not be considered for.